File: KGauss.java
import java.util.*;
import java.io.*;

public class KGauss
{
    public static double[] ngauss(double mu, double sigma, int n) {
        double[] data = new double[n];
        int i = 0;
        while(i < n) {
            double u = Math.random();
            double v = Math.random();
            data[i] = mu + sigma * Math.sqrt(-2 * Math.log(u)) 
                      * Math.cos(2 * Math.PI * v);
            i++;
            if(i < n) {
                data[i] = mu + sigma * Math.sqrt(-2 * Math.log(u)) 
                      * Math.sin(2 * Math.PI * v);
                i++;
            }
        }
        return data;
    }
    public static double ngauss(double mu, double sigma) {
        return ngauss(mu, sigma, 1)[0];
    }

    public static double[][] k_gauss(int k, int n, int dim, double lower,
                                           double upper, double scale) {
        double[][] data = new double[dim][n*k];
        int index = 0;
        for(int i = 0; i < k; i++) {
            double mu = Math.random() * (upper - lower) + lower;
            double sigma = Math.random() * scale * (upper - lower) 
                           + scale * lower;
            for(int j = 0; j < n; j++) {
                for(int d = 0; d < dim; d++) {
                    data[d][index] = ngauss(mu, sigma);
                }
                index++;
            }
        }
        return data;
    }

    private static String arrayToString(double[][] arr) {
        String s = "";
        for(double[] da : arr) {
            for(double d : da)
                s += Double.toString(d) + " ";
            if(da != arr[arr.length - 1])
                s += "\n";
        }
        return s;
    }

    public static String exportFile(String filename, double[][] data) {
        File outFile = new File(filename);
        BufferedWriter writer = null;
        try
        {
            writer = new BufferedWriter(new FileWriter(outFile));
            writer.write(arrayToString(data));
        }
        catch ( IOException e)
        {
            System.out.println("Could not write to " + filename);
        }
        finally
        {
            try
            {
                if ( writer != null)
                    writer.close();
            }
            catch ( IOException e)
            {
                System.out.println("Could not close writer for " + filename);
            }
         }
         return filename;
    }
    public static void main(String args[]) {
        int k = 2;
        int n = 1000;
        int dim = 2;
        int lower = -100;
        int upper = 100;
        double scale = 0.1;
        if(args.length > 1)
            k = Integer.parseInt(args[0]);
        if(args.length > 2)
            n = Integer.parseInt(args[1]);
        if(args.length > 3)
            dim = Integer.parseInt(args[2]);
        if(args.length > 4)
            lower = Integer.parseInt(args[3]);
        if(args.length > 5)
            upper = Integer.parseInt(args[4]);
        if(args.length > 6)
            scale = Double.parseDouble(args[5]);
        double[][] kg = k_gauss(k, n, dim, lower, upper, scale);
        exportFile("temp.txt", kg);
    }
}

File: kgauss.py
"""
A program to generate points from k gaussian distributions.
Author: Tom Shen
Date: 2013/01/29

Note:
Could also use numpy.random.multivariate_normal(mean, cov)
"""

from random import gauss, random
from math import pi
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import gaussian_kde
from scipy.stats import normaltest
from util import importFile, exportFile

def mvgauss(mean, cov, n):
    return np.random.multivariate_normal(mean, cov, (n))

def kmvgauss(k, n, cov, dim, lower=-100, upper=100):
    data = np.empty((k * n, dim))
    index = 0
    index = 0
    means = []
    for i in xrange(k):
        mean = []
        for d in xrange(dim):
            mu = random() * (upper - lower) + lower
            mean.append(mu)
        means.append(mean)
        dist = mvgauss(mean, cov, n)
        for p in dist:
            for d in xrange(dim):
                data[index][d] = p[d]
            index += 1
    return data, means

"""
My implementation of a Gaussian distribution generator, based on the 
Box-Muller transform.

returns an array of n random values
"""
def ngauss(mu, sigma, n=1):
    data = np.empty(n)
    i = 0
    while i < n:
        u = random()
        v = random()
        data[i] = mu + sigma * np.sqrt(-2 * np.log(u)) * np.cos(2 * pi * v)
        i += 1
        if i < n:
            data[i] = mu + sigma * np.sqrt(-2 * np.log(u)) * np.sin(2 * pi * v)
            i += 1
    return data

"""
k distributions
n points of data each
dim dimensional points
lower <= mu < upper

returns a list of point tuples
"""
def kgauss(k, n, dim=2, lower=-90, upper=90, sigma=3):
    data = np.empty((dim, k * n))
    index = 0
    for i in xrange(k):
        mu = random() * (upper - lower) + lower
        print mu
        for j in xrange(n):
            for d in xrange(dim):
                data[d][index] = ngauss(mu, sigma)
            index += 1
    return data

def kgauss_with_mus(k, n, dim=2, lower=-90, upper=90, sigma=3):
    data = np.empty((dim, k * n))
    index = 0
    mus = []
    for i in xrange(k):
        mu = random() * (upper - lower) + lower
        mus.append(mu)
        for j in xrange(n):
            for d in xrange(dim):
                data[d][index] = ngauss(mu, sigma)
            index += 1
    return data, mus
# random sigma somewhere between 0 and 2 * sigma
def kgauss_with_mus_sigmas(k, n, dim=2, lower=-90, upper=90, sigma=10):
    data = np.empty((dim, k * n))
    index = 0
    mus = []
    sigmas = []
    for i in xrange(k):
        mu = random() * (upper - lower) + lower
        mus.append(mu)
        sigma = random() * sigma + sigma
        sigmas.append(sigma)
        for j in xrange(n):
            for d in xrange(dim):
                data[d][index] = ngauss(mu, sigma)
            index += 1
    return data, mus, sigmas

def scatterPlot(data):
    plt.xlabel('x-coordinates')
    plt.ylabel('y-coordinates')
    plt.title('Toy Data for 2 gaussian distributions')
    plt.plot(data[0], data[1], 'ro')
    plt.show()

def pdf(data, bins, name):
    plt.xlabel(name)
    plt.ylabel('probability')
    plt.title('Probabilty Density Function')
    plt.hist(data, bins, normed=1)
    plt.show()

def cdf(data, name):
    plt.xlabel(name)
    plt.ylabel('cumulative probability')
    plt.title('Cumulative Probability Function')
    sorted_data = np.sort(data)
    plot_data = np.empty((2, sorted_data.size))
    num_entries = 0
    for i in xrange(sorted_data.size):
        num_entries += 1
        plot_data[0][i] = sorted_data[i]
        print plot_data[0][i]
        plot_data[1][i] = num_entries
    for i in xrange(plot_data[1].size):
        plot_data[1][i] /= num_entries
    plt.plot(plot_data[0], plot_data[1])
    plt.show()

def main():
    filename = 'temp.txt'
    data, means = kmvgauss(3, 100, [[10,50],[50,10]], 2)
    print data
    data = [[z[i] for z in data] for i in xrange(2)]
    plt.plot(data[0], data[1], 'ro')
    plt.show()
    #kg = kgauss(2, 100, lower=-90, upper=90)
    #exportFile(filename, kg)
    #kg = importFile(filename)

    #print normaltest(kg[0])
    #scatterPlot(kg)
    # pdf(kg[0], 40, 'x-coordinates')
    # cdf(kg[0], 'x-coordinates')
    
if __name__ == "__main__":
    main()